File: app/Models/MarkdownList.swift
import Foundation

/// What Return does on a line that is already part of a list or a quote.
///
/// Pure, so every rule here is tested without a text view. The caller supplies whether the
/// line is inside a fenced block - `MarkdownHighlighter` already keeps that answer, and a
/// `- ` inside ``` is code, not a bullet.
enum MarkdownList {
    enum Marker: Equatable, Sendable {
        case bullet(Character)
        case ordered(Int, delimiter: Character)

        /// What the *next* item's marker reads as. Ordered lists increment rather than
        /// renumbering what follows: renumbering rewrites lines nobody touched, makes one
        /// undo step span the whole list, and CommonMark renders `1. 1. 1.` correctly anyway.
        var next: String {
            switch self {
            case .bullet(let character): return "\(character) "
            case .ordered(let number, let delimiter): return "\(number + 1)\(delimiter) "
            }
        }
    }

    /// Everything before an item's text, kept verbatim so a continuation lines up under it.
    struct Prefix: Equatable, Sendable {
        /// Leading spaces, exactly as written.
        let indent: String
        /// The `> ` runs, exactly as written. A quote can carry a list and vice versa.
        let quotes: String
        let marker: Marker?
        /// Whether the item carried a `[ ]`, `[x]` or `[*]` box.
        let isTask: Bool
        /// UTF-16 offset where the item's own text starts.
        let contentStart: Int
        /// Nothing but whitespace after the marker.
        let isEmpty: Bool

        /// What to open the next line with. A task always continues unchecked - carrying
        /// `[x]` forward would tick a box nobody has done.
        var continuation: String {
            indent + quotes + (marker?.next ?? "") + (isTask ? "[ ] " : "")
        }

        /// Everything up to the item's text, which is what an empty item sheds.
        var length: Int { contentStart }
    }

    // MARK: - Parsing

    /// Where a `- `, `* `, `+ `, `1. ` or `1) ` marker ends. Shared with `MarkdownSyntax`,
    /// which needs the same answer to paint the marker, so the two cannot drift on what a
    /// bullet is.
    static func markerEnd(of body: Substring) -> String.Index? {
        guard let first = body.first else { return nil }

        if first == "-" || first == "*" || first == "+" {
            let next = body.index(after: body.startIndex)
            guard next < body.endIndex, body[next] == " " else { return nil }
            return body.index(after: next)
        }

        let digits = body.prefix { $0.isNumber }
        guard !digits.isEmpty, digits.count <= 9, digits.endIndex < body.endIndex,
              body[digits.endIndex] == "." || body[digits.endIndex] == ")" else { return nil }

        let after = body.index(after: digits.endIndex)
        guard after < body.endIndex, body[after] == " " else { return nil }
        return body.index(after: after)
    }

    /// Where a `[ ]`, `[x]` or `[*]` box ends, when one follows the marker.
    static func taskEnd(of rest: Substring) -> String.Index? {
        var index = rest.startIndex
        guard index < rest.endIndex, rest[index] == "[" else { return nil }

        index = rest.index(after: index)
        guard index < rest.endIndex, " xX*".contains(rest[index]) else { return nil }

        index = rest.index(after: index)
        guard index < rest.endIndex, rest[index] == "]" else { return nil }

        index = rest.index(after: index)
        guard index < rest.endIndex, rest[index] == " " else { return nil }
        return rest.index(after: index)
    }

    /// Nil when the line opens nothing that Return should carry forward.
    static func prefix(of line: String) -> Prefix? {
        var index = line.startIndex

        let indent = String(line[index...].prefix { $0 == " " })
        index = line.index(index, offsetBy: indent.count)

        var quotes = ""
        while index < line.endIndex, line[index] == ">" {
            var end = line.index(after: index)
            if end < line.endIndex, line[end] == " " { end = line.index(after: end) }
            quotes += line[index..<end]
            index = end
        }

        // Indentation after the quote bars belongs to the list, not to the quote.
        let inner = String(line[index...].prefix { $0 == " " })
        index = line.index(index, offsetBy: inner.count)

        var marker: Marker?
        var isTask = false

        if let end = markerEnd(of: line[index...]) {
            marker = parseMarker(line[index..<end])
            index = end
            if let boxed = taskEnd(of: line[index...]) {
                isTask = true
                index = boxed
            }
        }

        guard marker != nil || !quotes.isEmpty else { return nil }

        return Prefix(
            indent: indent + inner,
            quotes: quotes,
            marker: marker,
            isTask: isTask,
            contentStart: index.utf16Offset(in: line),
            isEmpty: line[index...].allSatisfy(\.isWhitespace)
        )
    }

    private static func parseMarker(_ token: Substring) -> Marker? {
        guard let first = token.first else { return nil }
        if first == "-" || first == "*" || first == "+" { return .bullet(first) }

        let digits = token.prefix { $0.isNumber }
        guard let number = Int(digits), digits.endIndex < token.endIndex else { return nil }
        return .ordered(number, delimiter: token[digits.endIndex])
    }

    // MARK: - Return

    enum Continuation: Equatable, Sendable {
        /// Let AppKit insert the newline it was going to.
        case none
        /// A newline plus the reconstructed prefix.
        case insert(String)
        /// An empty item sheds its marker instead of growing another one - the standard way
        /// out of a list. The range is relative to the line.
        case clear(NSRange)
    }

    /// - Parameters:
    ///   - caretColumn: UTF-16 offset of the caret within `line`.
    ///   - insideFence: a `- ` inside ``` is code, and Return there is just a newline.
    static func continuation(for line: String, caretColumn: Int, insideFence: Bool) -> Continuation {
        guard !insideFence, let prefix = prefix(of: line) else { return .none }

        // Return with the caret still inside the marker splits the line rather than
        // continuing it - there is no item yet to continue.
        guard caretColumn >= prefix.contentStart else { return .none }

        if prefix.isEmpty {
            // Only when the caret is at the end of it; mid-marker is the case above.
            return .clear(NSRange(location: 0, length: prefix.length))
        }
        return .insert("\n" + prefix.continuation)
    }
}

File: src/models/markdownList.ts
// What Return does on a line that is already part of a list or a quote. Pure, so every
// rule here is tested without an editor. The caller supplies whether the line is inside a
// fenced block - the highlighter already keeps that answer, and a `- ` inside ``` is code,
// not a bullet.

export type Marker =
  | { type: 'bullet'; char: string }
  | { type: 'ordered'; number: number; delimiter: string }

/** What the *next* item's marker reads as. Ordered lists increment rather than renumbering
 *  what follows: renumbering rewrites lines nobody touched, makes one undo step span the
 *  whole list, and CommonMark renders `1. 1. 1.` correctly anyway. */
export function nextMarker(marker: Marker): string {
  return marker.type === 'bullet' ? `${marker.char} ` : `${marker.number + 1}${marker.delimiter} `
}

/** Everything before an item's text, kept verbatim so a continuation lines up under it. */
export interface Prefix {
  /** Leading spaces, exactly as written. */
  indent: string
  /** The `> ` runs, exactly as written. A quote can carry a list and vice versa. */
  quotes: string
  marker: Marker | null
  /** Whether the item carried a `[ ]`, `[x]` or `[*]` box. */
  isTask: boolean
  /** UTF-16 offset where the item's own text starts. */
  contentStart: number
  /** Nothing but whitespace after the marker. */
  isEmpty: boolean
}

/** What to open the next line with. A task always continues unchecked - carrying `[x]`
 *  forward would tick a box nobody has done. */
export function continuationOf(prefix: Prefix): string {
  return prefix.indent + prefix.quotes + (prefix.marker ? nextMarker(prefix.marker) : '') + (prefix.isTask ? '[ ] ' : '')
}

/** Where a `- `, `* `, `+ `, `1. ` or `1) ` marker starting at `start` ends, or null.
 *  Shared with the syntax scanner, which needs the same answer to paint the marker. */
export function markerEnd(text: string, start: number): number | null {
  const first = text[start]
  if (first === undefined) return null
  if (first === '-' || first === '*' || first === '+') {
    return text[start + 1] === ' ' ? start + 2 : null
  }
  let digits = 0
  while (/[0-9]/.test(text[start + digits] ?? '')) digits++
  if (digits === 0 || digits > 9) return null
  const delimiter = text[start + digits]
  if (delimiter !== '.' && delimiter !== ')') return null
  return text[start + digits + 1] === ' ' ? start + digits + 2 : null
}

/** Where a `[ ]`, `[x]` or `[*]` box starting at `start` ends, when one follows the marker. */
export function taskEnd(text: string, start: number): number | null {
  if (text[start] !== '[') return null
  if (!' xX*'.includes(text[start + 1] ?? '\0')) return null
  if (text[start + 2] !== ']') return null
  if (text[start + 3] !== ' ') return null
  return start + 4
}

/** Null when the line opens nothing that Return should carry forward. */
export function prefixOf(line: string): Prefix | null {
  let index = 0
  while (line[index] === ' ') index++
  const indent = line.slice(0, index)

  let quotes = ''
  while (line[index] === '>') {
    let end = index + 1
    if (line[end] === ' ') end++
    quotes += line.slice(index, end)
    index = end
  }

  // Indentation after the quote bars belongs to the list, not to the quote.
  const innerStart = index
  while (line[index] === ' ') index++
  const inner = line.slice(innerStart, index)

  let marker: Marker | null = null
  let isTask = false
  const end = markerEnd(line, index)
  if (end !== null) {
    marker = parseMarker(line.slice(index, end))
    index = end
    const boxed = taskEnd(line, index)
    if (boxed !== null) {
      isTask = true
      index = boxed
    }
  }

  if (marker === null && quotes === '') return null

  return {
    indent: indent + inner,
    quotes,
    marker,
    isTask,
    contentStart: index,
    isEmpty: /^\s*$/.test(line.slice(index)),
  }
}

function parseMarker(token: string): Marker | null {
  const first = token[0]
  if (first === '-' || first === '*' || first === '+') return { type: 'bullet', char: first }
  const m = /^([0-9]+)([.)])/.exec(token)
  return m ? { type: 'ordered', number: Number(m[1]), delimiter: m[2] } : null
}

export type Continuation =
  /** Let the editor insert the newline it was going to. */
  | { type: 'none' }
  /** A newline plus the reconstructed prefix. */
  | { type: 'insert'; text: string }
  /** An empty item sheds its marker instead of growing another one - the standard way out
   *  of a list. `length` counts from the start of the line. */
  | { type: 'clear'; length: number }

/** caretColumn is the UTF-16 offset of the caret within `line`; insideFence says a `- `
 *  there is code, and Return is just a newline. */
export function continuation(line: string, caretColumn: number, insideFence: boolean): Continuation {
  if (insideFence) return { type: 'none' }
  const prefix = prefixOf(line)
  if (!prefix) return { type: 'none' }
  // Return with the caret still inside the marker splits the line rather than continuing
  // it - there is no item yet to continue.
  if (caretColumn < prefix.contentStart) return { type: 'none' }
  if (prefix.isEmpty) return { type: 'clear', length: prefix.contentStart }
  return { type: 'insert', text: '\n' + continuationOf(prefix) }
}
